webapp/src/ArchivalPage.js

```diff
@@ -1,19 +1,164 @@
-
+import { Input, Tag, Table, Timeline, Progress, Button, Space} from "antd";
+import { CheckOutlined, SyncOutlined,  } from '@ant-design/icons';
+import * as Stomp from '@stomp/stompjs';
+import { useMutex } from 'react-context-mutex';
 
 import * as API from "./api";
 import Header from "./Header";
-import Footer from "./Footer";
-
+import cloneDeep from 'lodash/cloneDeep';
 
 
+let id =  Math.floor(Math.random() * 1000);
 function ArchivalPage() {
     const [userData, setUserData] = useState(null);
     const [archivalSubscriptions, setArchivalSubscriptions] = useState([]);
     const [timelineEvents, setTimelineEvents] = useState([]);
-    const [videosInProgress, setVideoInProgress] = useState([]);
+    const [videoInProgressDataset, setVideoInProgressDataset] = useState([]);
+    const [conn, setConn] = useState(null);
+    const latest = useRef(videoInProgressDataset);
+
+    const MutexRunner = useMutex();
+    const mutex = new MutexRunner('messageHandler');
+
+    // TODO: currently connects every time the videos in progress changes
+    useEffect(()=> {
+            var client = new Stomp.Client({
+                // brokerURL: "ws://localhost:61614/ws",
+                 webSocketFactory: function () {
+                  return new WebSocket("ws://localhost/ws");
+                },
+                  connectHeaders: {
+                    login: 'guest', // TODO
+                    passcode: 'guest',
+                  },
+                reconnectDelay: 5000,
+                heartbeatIncoming: 4000,
+                heartbeatOutgoing: 4000,
+                connectionTimeout: 5000,
+              });
+
+              client.onConnect = function(frame) {
+                setConn(client);
+              };
+
+              client.onDisconnect = function(frame) {
+                setConn(null);
+              }
+
+              client.onWebSocketError = async (error: any) => {
+                console.log(`onWebSocketError ${JSON.stringify(error)}`, 'WS');
+            };
+              
+              client.onStompError = function (frame) {
+                // Will be invoked in case of error encountered at Broker
+                // Bad login/passcode typically will cause an error
+                // Complaint brokers will set `message` header with a brief message. Body may contain details.
+                // Compliant brokers will terminate the connection after any error
+                console.log('Broker reported error: ' + frame.headers['message']);
+                console.log('Additional details: ' + frame.body);
+              };
+              client.activate();
+
+              return () => client.deactivate();
+    
+    }, []);
+
+    function processMessage(message) {
+      mutex.lock();
+      setVideoInProgressDataset(videosInProg => {
+          if (videosInProg == null) {
+              return videosInProg;
+          }
+        let dataset = JSON.parse(JSON.stringify(videosInProg))
```

Comment:
Seems redundant

Reply:
it's a very stupid deep copy mechanism. I might be able to take it out now that I'm using the functional approach to setting this variable.